### PR TITLE
Добавить вес вертикалям.

### PR DESCRIPTION
**Pull request by polesye**
**Ticket:** [CORER-73](http://jira.npoed.ru/browse/CORER-73)

**Docs:**
- http://confluence.npoed.ru/pages/viewpage.action?pageId=1901252
- http://confluence.npoed.ru/pages/viewpage.action?pageId=1901298

**sandbox:**
http://rnoepgrading.raccoongang.com
http://rnoepgrading.raccoongang.com:18010

@auraz please review.

@den99f FYI

### CODE REVIEW

**Comment by polesye** [comment]
@auraz Where is your review?

**Comment by auraz** [comment]
@polesye sorry, in the progress.

File: common/lib/xmodule/xmodule/tests/test_graders.py [hunk 1]
```diff
@@ -38,15 +39,32 @@ def test_weighted_grading(self):
...
-
-    empty_gradesheet = {
-    }
-
-    incomplete_gradesheet = {
-        'Homework': [],
-        'Lab': [],
-        'Midterm': [],
-    }
+    def setUp(self):
```
Comment (auraz) [inline common/lib/xmodule/xmodule/tests/test_graders.py]: сan override_settings work here?

File: lms/envs/common.py [hunk 2]
```diff
@@ -2578,8 +2578,9 @@
...
 # Possible values: vertical|sequential
 GRADING_TYPE = 'vertical'
-GRADING_ALLOWED_TYPES = ('vertical', 'sequential')
-
 # If grading type is set to `vertical`, we can adjust progress view to display
 # only gradable verticals.
 DISPLAY_ONLY_GRADABLE_VERTICALS = True
+
+COURSE_GRADER = 'WeightedSubsectionsGrader'
+ASSIGNMENT_GRADER = 'WeightedAssignmentFormatGrader'
```
Comment (auraz) [inline lms/envs/common.py]: please document graders here: what they do

File: lms/envs/test.py [hunk 3]
```diff
@@ -494,6 +494,9 @@
 FEATURES['ENABLE_LTI_PROVIDER'] = True
 INSTALLED_APPS += ('lti_provider',)
 AUTHENTICATION_BACKENDS += ('lti_provider.users.LtiBackend',)
+
+######## Grading ###########
 FEATURES['ENABLE_CUSTOM_GRADING'] = False
 GRADING_TYPE = 'vertical'
-GRADING_ALLOWED_TYPES = ('vertical', 'sequential')
+
+ASSIGNMENT_GRADER = 'AssignmentFormatGrader'
```
Comment (auraz) [inline lms/envs/test.py]: please document graders here: what they do and why we use it here

File: openedx/core/djangoapps/grading_policy/graders/assignment_format.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
+""""
+The file contains AssignmentFormatGrader.
+"""
```
Comment (auraz) [inline openedx/core/djangoapps/grading_policy/graders/assignment_format.py]: update module docstring with high level grader description

File: openedx/core/djangoapps/grading_policy/graders/single_section.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+""""
+The file contains SingleSectionGrader.
```
Comment (auraz) [inline openedx/core/djangoapps/grading_policy/graders/single_section.py]: add docs here too plz

File: openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
+""""
+The file contains WeightedAssignmentFormatGrader.
+"""
```
Comment (auraz) [inline openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py]: add docs here too plz

File: openedx/core/djangoapps/grading_policy/graders/weighted_subs.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
+""""
+The file contains WeightedSubsectionsGrader.
+"""
```
Comment (auraz) [inline openedx/core/djangoapps/grading_policy/graders/weighted_subs.py]: add docs here too plz

File: openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    displayed, scores from the same category will be similar (for example, by color).
+
+    section_type is a string that is the type of a singular section. For example, for Labs it
+    would be "Lab". This defaults to be the same as category.
+
+    short_label is similar to section_type, but shorter. For example, for Homework it would be
+    "HW".
+
+    starting_index is the first number that will appear. For example, starting_index=3 and
+    min_count = 2 would produce the labels "Assignment 3", "Assignment 4"
```
Comment (auraz) [inline openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py]: check this docs for correctness

File: openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        self.hide_average = hide_average
+
+    def grade(self, grade_sheet, generate_random_scores=False):
+        def total_with_drops(breakdown, drop_count):
+            '''calculates total score for a section while dropping lowest scores'''
+            # A list of the indices of the dropped scores
+            dropped_indices = []
+            dropped_breakdown = []
+            if drop_count > 0:
+                # create an array of tuples with (index, mark), sorted by mark['percent'] descending
```
Comment (auraz) [inline openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py]: please fix comment ( * weighted)

File: openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                sorted_breakdown = sorted(enumerate(breakdown), key=lambda x: -x[1]['percent'] * x[1]['weight'])
+                dropped_breakdown = [x[1] for x in sorted_breakdown[-drop_count:]]
+                dropped_indices = [x[0] for x in sorted_breakdown[-drop_count:]]
+
+            normalized_weight = sum(x['weight'] for x in breakdown) - sum(x['weight'] for x in dropped_breakdown)
+            aggregate_score = sum(m['percent'] * m['weight'] for m in breakdown if m not in dropped_breakdown)
+            if len(breakdown) - drop_count > 0 and normalized_weight > 0:
+                aggregate_score /= normalized_weight
+            elif normalized_weight == 0:
```
Comment (auraz) [inline openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py]: better else

File: openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+                aggregate_score = 1.0
+            return aggregate_score, dropped_indices
```
Comment (auraz) [inline openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py]: add assert for the rest of cases,

File: openedx/core/djangoapps/grading_policy/tests/test_ccx_views.py [hunk 12]
```diff
@@ -128,8 +146,8 @@ def setUp(self):
                     data=StringResponseXMLFactory().build_xml(answer='foo'),
                     metadata={'rerandomize': 'always'}
                 ) for _ in xrange(4)
-                ] for section in sections
-            ]
+            ] for section in sections
+        ]
```
Comment (auraz) [inline openedx/core/djangoapps/grading_policy/tests/test_ccx_views.py]: Please add CCX testing in test plan

File: openedx/core/djangoapps/grading_policy/tests/test_ccx_views.py [hunk 13]
```diff
@@ -111,14 +109,34 @@ def setUp(self):
...
             ItemFactory.create(
                 parent=chapter,
                 category="vertical",
-                metadata={'graded': True, 'format': 'Homework'})
-            for _ in xrange(4)]
+                metadata={'graded': True, 'format': 'Homework', 'weight': 0.5}
+            ),
+            ItemFactory.create(
+                parent=chapter,
+                category="vertical",
```
Comment (auraz) [inline openedx/core/djangoapps/grading_policy/tests/test_ccx_views.py]: please explain new course structure and why we do so and rename sections to vertical

File: openedx/core/djangoapps/grading_policy/tests/test_courseware_submitting_problems.py [hunk 14]
```diff
@@ -1,6 +1,8 @@
 # -*- coding: utf-8 -*-
 """
 lms/djangoapps/courseware/tests/test_submitting_problems.py
+python ./manage.py lms test --verbosity=1 \
+openedx/core/djangoapps/grading_policy/tests/test_courseware_submitting_problems.py --traceback --settings=test
```
Comment (auraz) [inline openedx/core/djangoapps/grading_policy/tests/test_courseware_submitting_problems.py]: Please add odcstring why this file exists.

File: openedx/core/djangoapps/grading_policy/vertical.py [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+
+
+# This is a tuple for holding scores from sections.
+# Section indicates the name of the section
+WeightedScore = namedtuple("WeightedScore", "earned possible graded section module_id weight")
+
+
+def aggregate_section_scores(scores, section_name="summary", weight=1.0):
+    """
+    scores: A list of WeightedScore objects
```
Comment (auraz) [inline openedx/core/djangoapps/grading_policy/vertical.py]: Fix docstring -> add description, what method does.

File: openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py (hunk 8 above)
Comment (polesye) [inline openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py]: > check this docs for correctness

checked. Correct.

File: lms/envs/common.py (hunk 2 above)
Comment (polesye) [inline lms/envs/common.py]: > please document graders here: what they do

Done

File: lms/envs/test.py (hunk 3 above)
Comment (polesye) [inline lms/envs/test.py]: > please document graders here: what they do and why we use it here

Done

File: openedx/core/djangoapps/grading_policy/graders/assignment_format.py (hunk 4 above)
Comment (polesye) [inline openedx/core/djangoapps/grading_policy/graders/assignment_format.py]: > update module docstring with high level grader description

Done

File: openedx/core/djangoapps/grading_policy/graders/single_section.py (hunk 5 above)
Comment (polesye) [inline openedx/core/djangoapps/grading_policy/graders/single_section.py]: > add docs here too plz

Done

File: openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py (hunk 6 above)
Comment (polesye) [inline openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py]: > add docs here too plz

Done

File: openedx/core/djangoapps/grading_policy/graders/weighted_subs.py (hunk 7 above)
Comment (polesye) [inline openedx/core/djangoapps/grading_policy/graders/weighted_subs.py]: > add docs here too plz

Done

File: openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py (hunk 9 above)
Comment (polesye) [inline openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py]: > please fix comment ( * weighted)

Done

File: openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py (hunk 11 above)
Comment (polesye) [inline openedx/core/djangoapps/grading_policy/graders/weighted_assignment_format.py]: > add assert for the rest of cases

added

File: openedx/core/djangoapps/grading_policy/tests/test_ccx_views.py (hunk 13 above)
Comment (polesye) [inline openedx/core/djangoapps/grading_policy/tests/test_ccx_views.py]: > please explain new course structure and why we do so and rename sections to vertical

Done

File: openedx/core/djangoapps/grading_policy/vertical.py (hunk 15 above)
Comment (polesye) [inline openedx/core/djangoapps/grading_policy/vertical.py]: > Fix docstring -> add description, what method does.

Done

File: openedx/core/djangoapps/grading_policy/tests/test_courseware_submitting_problems.py (hunk 14 above)
Comment (polesye) [inline openedx/core/djangoapps/grading_policy/tests/test_courseware_submitting_problems.py]: > Please add docstring why this file exists.

done

**Comment by polesye** [comment]
@auraz I will megre this PR tomorrow.

File: common/lib/xmodule/xmodule/tests/test_graders.py (hunk 1 above)
Comment (polesye) [inline common/lib/xmodule/xmodule/tests/test_graders.py]: > сan override_settings work here?

done.

**Pull request merged**